File: android_dependencies/src/main/java/com/app/dependencies/data/dao/RealmTextBookmarkDAO.kt
package com.app.dependencies.data.dao

import com.app.dependencies.data.dto.RealmTextBookmarkDTO
import com.app.dependencies.data.dto.RealmTopicDTO
import com.app.shared.data.dao.TextBookmarkDAO
import com.app.shared.data.dto.BookmarkDTO
import com.app.shared.data.dto.TopicDTO

class RealmTextBookmarkDAO(override val instance: RealmProvider): RealmDAO, TextBookmarkDAO {

    override fun getAll(): List<BookmarkDTO.TextBookmarkDTO> {
        val realm = instance()
        val result = realm.where(RealmTextBookmarkDTO::class.java).findAll()
        return realm.copyFromRealm(result).toList()
    }

    override fun insert(dto: BookmarkDTO.TextBookmarkDTO) {
        val realm = instance()

        val topicDTO = dto.topic?.let {
            RealmTopicDTO(id = it.id, name = it.name)
        }

        val roomDTO = RealmTextBookmarkDTO(id = dto.id, date = dto.date, text = dto.text, topic = topicDTO)

        realm.beginTransaction()
        realm.insertOrUpdate(roomDTO)
        realm.commitTransaction()
    }

    override fun delete(bookmarkId: Int) {
        val realm = instance()
        val result = realm.where(RealmTextBookmarkDTO::class.java).equalTo("id", bookmarkId).findAll()
        realm.beginTransaction()
        result.deleteAllFromRealm()
        realm.commitTransaction()
    }

    override fun replaceTopic(topicId: Int, withTopicDTO: TopicDTO) {
        val realm = instance()

        val topicResult = realm.where(RealmTopicDTO::class.java).equalTo("id", withTopicDTO.id).findFirst() ?: return
        val bookmarkResult = realm.where(RealmTextBookmarkDTO::class.java).equalTo("topic.id", topicId).findAll() ?: return

        realm.beginTransaction()

        bookmarkResult.forEach {
            it.topic = topicResult
        }

        realm.commitTransaction()
    }
}

File: ios_app/ios_dependencies/ios_dependencies/Data/DAO/RealmTextBookmarkDAO.swift
//
//  RealmTextBookmarkDAO.swift
//  ios_dependencies
//
//  Created by Gabriel Coman on 14/04/2020.
//  Copyright © 2020 Gabriel Coman. All rights reserved.
//

import Foundation
import RememberShared
import Realm
import RealmSwift

class RealmTextBookmarkDAO: RealmDAO, TextBookmarkDAO {
    
    let realm: Realm?
    
    public init(realm: Realm?){
        self.realm = realm
    }
    
    func insert(dto____ dto: BookmarkDTOTextBookmarkDTO) {
        try? realm?.write {
            realm?.add(dto.toObject(), update: Realm.UpdatePolicy.all)
        }
    }
    
    func getAll() -> [BookmarkDTOTextBookmarkDTO] {
        let result = realm?.objects(RealmTextBookmarkObject.self)
        return result?.map { $0 }.map { $0.toDTO() } ?? []
    }
    
    func delete(bookmarkId: Int32) {
        let result = realm?.objects(RealmTextBookmarkObject.self).filter { $0.id == bookmarkId }
        
        guard let dto = result?.first else {
            return
        }
        
        try? realm?.write {
            realm?.delete(dto)
        }
    }
    
    func replaceTopic(topicId: Int32, withTopicDTO: TopicDTO) {
        let topicResult = realm?.objects(RealmTopicObject.self).filter { $0.id == withTopicDTO.id }
        let bookmarkResult = realm?.objects(RealmTextBookmarkObject.self).filter { $0.topic?.id == topicId }
        
        guard let defaultTopic = topicResult?.first, let bookmarks = bookmarkResult else {
            return
        }
        
        try? realm?.write {
            bookmarks.forEach {
                $0.topic = defaultTopic
            }
        }
    }
}
